### PR TITLE
add service repository resource, with tests

### PR DESCRIPTION
**Pull request by davidbloss**
## Issues

https://github.com/OpsLevel/team-platform/issues/308

## Changelog

[service_repository public docs](https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/resources/service_repository)

Changes:
- Add `service_repository` resource, with tests

- [X] List your changes here
- [ ] Make a `changie` entry

### Using this file:
```tf
# main.tf
resource "opslevel_service_repository" "with_id" {
  repository     = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNzcxNzQ"
  service        = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8zMjc1NA"
  base_directory = "path"
  name           = "TEST"
}

resource "opslevel_service_repository" "with_alias" {
  service_alias    = "awesome_rails_service"
  repository_alias = "gitlab.com:test4605/monster-flowers-mail-book"
}
```

### Create Service Dependency, `terraform apply`
```tf
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # opslevel_service_repository.with_alias will be created
  + resource "opslevel_service_repository" "with_alias" {
      + id               = (known after apply)
      + last_updated     = (known after apply)
      + repository_alias = "gitlab.com:test4605/monster-flowers-mail-book"
      + service_alias    = "awesome_rails_service"
    }

  # opslevel_service_repository.with_id will be created
  + resource "opslevel_service_repository" "with_id" {
      + base_directory = "path"
      + id             = (known after apply)
      + last_updated   = (known after apply)
      + name           = "TEST"
      + repository     = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNzcxNzQ"
      + service        = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8zMjc1NA"
    }

Plan: 2 to add, 0 to change, 0 to destroy.
opslevel_service_repository.with_id: Creating...
opslevel_service_repository.with_alias: Creating...
opslevel_service_repository.with_alias: Creation complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzE]
opslevel_service_repository.with_id: Creation complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

### Update this file:
```tf
# main.tf
resource "opslevel_service_repository" "with_id" {
  repository     = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNzcxNzQ"
  service        = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8zMjc1NA"
  base_directory = "path/new"
  name           = "TEST2"
}

resource "opslevel_service_repository" "with_alias" {
  service_alias    = "awesome_rails_service"
  repository_alias = "gitlab.com:test4605/monster-flowers-mail-book"
}
```

### Update Service Dependency, `terraform apply`
```tf
opslevel_service_repository.with_id: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
opslevel_service_repository.with_alias: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzE]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_service_repository.with_id will be updated in-place
  ~ resource "opslevel_service_repository" "with_id" {
      ~ base_directory = "path" -> "path/new"
        id             = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI"
      ~ last_updated   = "Friday, 12-Apr-24 14:28:15 CDT" -> (known after apply)
      ~ name           = "TEST" -> "TEST2"
        # (2 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
opslevel_service_repository.with_id: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
opslevel_service_repository.with_id: Modifications complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

### Update this file, SHOWS KNOWN BUG:
```tf
# main.tf
resource "opslevel_service_repository" "with_id" {
  repository     = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNzcxNzQ"
  service        = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8zMjc1NA"
  base_directory = "path/new"
  # name           = "TEST2"  <-- cannot unset, known API bug
}

resource "opslevel_service_repository" "with_alias" {
  service_alias    = "awesome_rails_service"
  repository_alias = "gitlab.com:test4605/monster-flowers-mail-book"
}
``` 

### Update Service Dependency, `terraform apply`
```tf
opslevel_service_repository.with_id: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
opslevel_service_repository.with_alias: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzE]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_service_repository.with_id will be updated in-place
  ~ resource "opslevel_service_repository" "with_id" {
        id             = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI"
      ~ last_updated   = "Friday, 12-Apr-24 14:28:39 CDT" -> (known after apply)
      - name           = "TEST2" -> null
        # (3 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
opslevel_service_repository.with_id: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
╷
│ Error: Known error
│ 
│   with opslevel_service_repository.with_id,
│   on main.tf line 1, in resource "opslevel_service_repository" "with_id":
│    1: resource "opslevel_service_repository" "with_id" {
│ 
│ Unable to unset 'name' field for now. We have a planned fix for this.
```
☝️ this check runs before the API is called, so we won't have any weird data issues.

### Update config to unset `base_directory` - `terraform apply`
```tf
resource "opslevel_service_repository" "with_id" {
  repository = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNzcxNzQ"
  service    = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8zMjc1NA"
  # base_directory = "path/new"
  name = "TEST2"
}

resource "opslevel_service_repository" "with_alias" {
  service_alias    = "awesome_rails_service"
  repository_alias = "gitlab.com:test4605/monster-flowers-mail-book"
}
```

### Update Service Dependency, to unset `base_directory` - `terraform apply`
```tf
opslevel_service_repository.with_id: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
opslevel_service_repository.with_alias: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzE]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_service_repository.with_id will be updated in-place
  ~ resource "opslevel_service_repository" "with_id" {
      - base_directory = "path/new" -> null
        id             = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI"
      ~ last_updated   = "Friday, 12-Apr-24 14:28:39 CDT" -> (known after apply)
        name           = "TEST2"
        # (2 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
opslevel_service_repository.with_id: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
opslevel_service_repository.with_id: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

### Destroy Service Dependency, `terraform destroy`
```tf
opslevel_service_repository.with_id: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
opslevel_service_repository.with_alias: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzE]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # opslevel_service_repository.with_alias will be destroyed
  - resource "opslevel_service_repository" "with_alias" {
      - id               = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzE" -> null
      - last_updated     = "Friday, 12-Apr-24 14:28:15 CDT" -> null
      - repository_alias = "gitlab.com:test4605/monster-flowers-mail-book" -> null
      - service_alias    = "awesome_rails_service" -> null
    }

  # opslevel_service_repository.with_id will be destroyed
  - resource "opslevel_service_repository" "with_id" {
      - id           = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI" -> null
      - last_updated = "Friday, 12-Apr-24 14:31:59 CDT" -> null
      - name         = "TEST2" -> null
      - repository   = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNzcxNzQ" -> null
      - service      = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8zMjc1NA" -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.
opslevel_service_repository.with_id: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzI]
opslevel_service_repository.with_alias: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZVJlcG9zaXRvcnkvNjYxNzE]
opslevel_service_repository.with_alias: Destruction complete after 1s
opslevel_service_repository.with_id: Destruction complete after 1s

Destroy complete! Resources: 2 destroyed.
```

### CODE REVIEW

File: tests/resources.tf [hunk 1]
```diff
@@ -545,43 +601,3 @@ resource "opslevel_check_tool_usage" "example" {
...
-  category = var.test_id
-  level    = var.test_id
-  owner    = var.test_id
-  filter   = var.test_id
-  property = "language"
-  predicate = {
-    type  = "equals"
-    value = "python"
-  }
-}
```
Comment (rocktavious) [inline tests/resources.tf:587]: why are these deleted?

Comment (rocktavious) [reply tests/resources.tf:587]: oh did you alphabetize them?

File: opslevel/resource_opslevel_service_repository.go [hunk 2]
```diff
@@ -1,180 +1,295 @@
...
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &planModel)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var nameBeforeUpdate types.String
+	resp.Diagnostics.Append(req.State.GetAttribute(ctx, path.Root("name"), &nameBeforeUpdate)...)
+	if !nameBeforeUpdate.IsNull() && planModel.Name.IsNull() {
+		resp.Diagnostics.AddError("Known error", "Unable to unset 'name' field for now. We have a planned fix for this.")
```
Comment (rocktavious) [inline opslevel/resource_opslevel_service_repository.go:256]: nice touch

**Review by rocktavious — APPROVED** [review]
LGTM

File: tests/resources.tf (hunk 1 above)
Comment (davidbloss) [reply tests/resources.tf:587]: yes 😁
